src/main/host/descriptor/timerfd.c

```diff
@@ -61,28 +100,50 @@ static void _timerfd_close(LegacyDescriptor* descriptor, Host* host) {
 }
 
 static void _timerfd_free(LegacyDescriptor* descriptor) {
-    TimerFd* timer = _timerfd_fromLegacyDescriptor(descriptor);
-    MAGIC_ASSERT(timer);
-    descriptor_clear((LegacyDescriptor*)timer);
-    MAGIC_CLEAR(timer);
-    g_free(timer);
+    TimerFd* timerfd = _timerfd_fromLegacyDescriptor(descriptor);
+    MAGIC_ASSERT(timerfd);
+    descriptor_clear((LegacyDescriptor*)timerfd);
+    if (timerfd->timer) {
+        timer_unref(timerfd->timer);
+        timerfd->timer = NULL;
+    }
+    MAGIC_CLEAR(timerfd);
+    g_free(timerfd);
     worker_count_deallocation(TimerFd);
 }
 
+static void _timerfd_cleanup(LegacyDescriptor* descriptor) {
+    TimerFd* timerfd = _timerfd_fromLegacyDescriptor(descriptor);
+    MAGIC_ASSERT(timerfd);
+
+    if (timerfd->timer) {
+        // Break circular reference; the timer has a task with a reference to
+        // this descriptor.
+        timer_unref(timerfd->timer);
+        timerfd->timer = NULL;
+    }
+}
+
 static DescriptorFunctionTable _timerfdFunctions = {
-    _timerfd_close, NULL, _timerfd_free, MAGIC_VALUE};
+    _timerfd_close, _timerfd_cleanup, _timerfd_free, MAGIC_VALUE};
+
+static void _timerfd_expire(Host* host, gpointer voidTimer, gpointer data);
 
 TimerFd* timerfd_new() {
-    TimerFd* timer = g_new0(TimerFd, 1);
-    MAGIC_INIT(timer);
-    MAGIC_INIT(&timer->timer);
+    TimerFd* timerfd = g_new0(TimerFd, 1);
+    MAGIC_INIT(timerfd);
 
-    descriptor_init(&(timer->super), DT_TIMER, &_timerfdFunctions);
-    descriptor_adjustStatus(&(timer->super), STATUS_DESCRIPTOR_ACTIVE, TRUE);
+    descriptor_init(&(timerfd->super), DT_TIMER, &_timerfdFunctions);
+    descriptor_adjustStatus(&(timerfd->super), STATUS_DESCRIPTOR_ACTIVE, TRUE);
+
+    descriptor_refWeak(timerfd);
+    Task* task = task_new(_timerfd_expire, timerfd, NULL, descriptor_unrefWeak, NULL);
+    timerfd->timer = timer_new(task);
```

Comment:
Why is this task needed when it wasn't before? (Before this PR there was only a single `task_new()` call in `timer.c`, but this PR has a `task_new()` call in both `timerfd.c` and `timer.c`).

Reply:
Before there was a single task object that both did the timer stuff (check whether it had been cancelled, whether it should fire now, whether it should reschedule, etc) and the timerfd stuff (notify descriptor listeners).

Now there's a Timer task that does the timer stuff (and whatever extra task you hand it), and a TimerFd task that notifies descriptor listeners